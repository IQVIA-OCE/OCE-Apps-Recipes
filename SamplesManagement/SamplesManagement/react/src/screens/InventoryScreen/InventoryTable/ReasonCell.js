import React, { useContext, useEffect } from 'react';
import { Select, Text } from 'apollo-react-native';
import { useField, useFormikContext } from 'formik';
import { InventoryContext } from '../InventoryContext';
import { INVENTORY_STATUS, REASON_LIST } from '../constants';
import { INVENTORY_FORM_TYPE } from '../../../constants/Inventories';
import { StyleSheet, View } from 'react-native';

const ReasonCell = ({ row }) => {
  const { values } = useFormikContext();
  const {
    editingType,
    config: { showCalculatedFields, showSystemCount },
  } = useContext(InventoryContext);

  const fieldIndex = values.products.findIndex(
    ({ lotNumberId }) => row.lotNumberId === lotNumberId
  );
  const [field, { touched, error }, helper] = useField(
    `products[${fieldIndex}]`
  );

  useEffect(() => {
    if (editingType !== INVENTORY_FORM_TYPE.preview) {
      if (parseInt(row.physicalQuantity) === parseInt(row.systemCount)) {
        helper.setValue({
          ...field.value,
          discrepancyReason: '',
          reasonObj: null,
        });
      }
      if (field.value.discrepancyReason && !field.value.reasonObj) {
        helper.setValue({
          ...field.value,
          reasonObj: REASON_LIST.find(
            el => el.id === field.value.discrepancyReason
          ),
        });
      }
    }
  }, [row.physicalQuantity]);

  // custom logic on specific configuration
  if (
    editingType === INVENTORY_FORM_TYPE.edit &&
    showCalculatedFields &&
    !showSystemCount &&
    values.status !== INVENTORY_STATUS.saved
  ) {
    return null;
  }

  if (
    (editingType === INVENTORY_FORM_TYPE.edit ||
      editingType === INVENTORY_FORM_TYPE.editSaved) &&
    !row.selected &&
    parseInt(row.physicalQuantity) === parseInt(row.systemCount)
  ) {
    return null;
  }

  if (editingType === INVENTORY_FORM_TYPE.preview) {
    return <Text>{field.value.discrepancyReason}</Text>;
  }

  return (
    <View style={styles.root}>
      <Select
        options={REASON_LIST}
        placeholder={'-None-'}
        value={field.value.reasonObj}
        onChange={value => {
          helper.setValue({
            ...field.value,
            discrepancyReason: value ? value.id : '',
            reasonObj: value,
          });
        }}
        fullWidth
        error={
          touched && error && error.discrepancyReason
            ? error.discrepancyReason
            : null
        }
        helperText={
          touched && error && error.discrepancyReason
            ? error.discrepancyReason
            : null
        }
      />
    </View>
  );
};

const styles = StyleSheet.create({
  root: {
    paddingVertical: 5,
    width: '100%',
  },
});

export default ReasonCell;
