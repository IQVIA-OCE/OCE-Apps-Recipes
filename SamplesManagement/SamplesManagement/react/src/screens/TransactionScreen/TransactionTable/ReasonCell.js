import React, { useEffect } from 'react';
import { StyleSheet, View } from 'react-native';
import { useField } from 'formik';
import { TextInput, Select } from 'apollo-react-native';


const ReasonCell = ({ form, row, ...rest }) => {
  const fieldIndex = form.values.products.findIndex(({ Id }) => row.Id === Id);
  const [reasonField, reasonMeta, reasonHelpers] = useField(
    `products[${fieldIndex}].reason`
  );
  return (
    <View style={styles.container}>
      <Select
        options={[
          { label: 'Theft/Loss', id: '1' },
          { label: 'Incorrect Disbursement', id: '2' },
          { label: 'Other', id: '3' },
        ]}
        placeholder={'-None-'}
        value={reasonField.value}
        onChange={value => {
          reasonHelpers.setValue(value);
        }}
        style={{ width: '100%' }}
        fullWidth
        hideDropdownPlaceholder
        error={form.touched.products && form.touched.products.length && reasonMeta.error}
        helperText={form.touched.products && form.touched.products.length ? reasonMeta.error : null}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    width: '100%'
  },
  input: {
    width: 80,
  },
});

export default ReasonCell;
