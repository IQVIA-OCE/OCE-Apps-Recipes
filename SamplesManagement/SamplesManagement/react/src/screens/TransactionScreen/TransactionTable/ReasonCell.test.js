import React from 'react';
import renderer, { act } from 'react-test-renderer';
import ReasonCell from './ReasonCell';
import { FieldArray, Formik } from 'formik';
import { Select } from 'apollo-react-native';

const Component = () => (
  <Formik initialValues={{ products: [{ Id: '1', comments: '' }] }}>
    {() => (
      <FieldArray name="products">
        {props => <ReasonCell row={{ Id: '1' }} {...props} />}
      </FieldArray>
    )}
  </Formik>
);

describe('ReasonCell', () => {
  it('should render properly', async () => {
    let tree;
    act(() => {
      tree = renderer.create(<Component />);
    });

    act(() => {
      tree.root.findByType(Select).props.onChange({ id: '1', label: '1' });
    });

    expect(tree.toJSON()).toMatchSnapshot();
    act(() => {
      tree.update(<Component />);
    });

    expect(tree.toJSON()).toMatchSnapshot();
    act(() => {
      tree.update();
    });
  });
});
