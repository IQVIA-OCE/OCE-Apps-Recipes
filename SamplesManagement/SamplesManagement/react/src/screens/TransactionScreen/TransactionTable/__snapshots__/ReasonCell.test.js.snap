// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReasonCell should render properly 1`] = `
<View
  style={
    Object {
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "width": "100%",
        },
        Object {
          "width": "100%",
        },
      ]
    }
  >
    <View>
      <View
        collapsable={false}
      >
        <View
          onLayout={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              },
              Object {
                "backgroundColor": undefined,
                "borderColor": "#d9d9d9",
                "borderRadius": 4,
                "borderWidth": 1,
                "elevation": 0,
                "height": 40,
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "stretch",
                "flex": 1,
                "flexDirection": "row",
                "flexGrow": 1,
                "paddingLeft": 8,
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              hitSlop={
                Object {
                  "bottom": 14,
                  "left": 14,
                  "right": 14,
                  "top": 14,
                }
              }
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "flex": 1,
                  "opacity": 1,
                }
              }
            >
              <Text
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontSize": 16,
                      "fontWeight": "400",
                    },
                  ]
                }
              >
                1
              </Text>
            </View>
          </View>
          <View
            accessibilityComponentType="button"
            accessibilityRole="button"
            accessibilityStates={Array []}
            accessibilityTraits="button"
            accessible={true}
            centered={true}
            focusable={true}
            hitSlop={
              Object {
                "bottom": 6,
                "left": 6,
                "right": 6,
                "top": 6,
              }
            }
            onClick={[Function]}
            onLayout={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            pointerEvents="box-only"
          >
            <View
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "alignItems": "center",
                      "justifyContent": "center",
                      "margin": 6,
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 18,
                      "height": 36,
                      "width": 36,
                    },
                    undefined,
                    Object {
                      "margin": 0,
                    },
                  ],
                ]
              }
            >
              <View>
                <Text
                  accessibilityElementsHidden={true}
                  allowFontScaling={false}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontSize": 24,
                      },
                      Array [
                        Object {
                          "transform": Array [
                            Object {
                              "scaleX": 1,
                            },
                          ],
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                      ],
                      Object {
                        "fontFamily": "Material Design Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  ï…€
                </Text>
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "bottom": 0,
                    "left": 0,
                    "overflow": "hidden",
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  Object {
                    "borderRadius": 0,
                  },
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`ReasonCell should render properly 2`] = `
<View
  style={
    Object {
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "width": "100%",
        },
        Object {
          "width": "100%",
        },
      ]
    }
  >
    <View>
      <View
        collapsable={false}
      >
        <View
          onLayout={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              },
              Object {
                "backgroundColor": undefined,
                "borderColor": "#d9d9d9",
                "borderRadius": 4,
                "borderWidth": 1,
                "elevation": 0,
                "height": 40,
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "stretch",
                "flex": 1,
                "flexDirection": "row",
                "flexGrow": 1,
                "paddingLeft": 8,
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              hitSlop={
                Object {
                  "bottom": 14,
                  "left": 14,
                  "right": 14,
                  "top": 14,
                }
              }
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "flex": 1,
                  "opacity": 1,
                }
              }
            >
              <Text
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontSize": 16,
                      "fontWeight": "400",
                    },
                  ]
                }
              >
                1
              </Text>
            </View>
          </View>
          <View
            accessibilityComponentType="button"
            accessibilityRole="button"
            accessibilityStates={Array []}
            accessibilityTraits="button"
            accessible={true}
            centered={true}
            focusable={true}
            hitSlop={
              Object {
                "bottom": 6,
                "left": 6,
                "right": 6,
                "top": 6,
              }
            }
            onClick={[Function]}
            onLayout={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            pointerEvents="box-only"
          >
            <View
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "alignItems": "center",
                      "justifyContent": "center",
                      "margin": 6,
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 18,
                      "height": 36,
                      "width": 36,
                    },
                    undefined,
                    Object {
                      "margin": 0,
                    },
                  ],
                ]
              }
            >
              <View>
                <Text
                  accessibilityElementsHidden={true}
                  allowFontScaling={false}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontSize": 24,
                      },
                      Array [
                        Object {
                          "transform": Array [
                            Object {
                              "scaleX": 1,
                            },
                          ],
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                      ],
                      Object {
                        "fontFamily": "Material Design Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  ï…€
                </Text>
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "bottom": 0,
                    "left": 0,
                    "overflow": "hidden",
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  Object {
                    "borderRadius": 0,
                  },
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
